### PR TITLE
Best efforts recovery recover seqno prefix

### PR DESCRIPTION
**Pull request by jowlyzhang**
This PR make best efforts recovery more permissive by allowing it to recover incomplete Version that presents a valid point in time view from the user's perspective. Currently, a Version is only valid and saved if all files consisting that Version can be found. With this change, if only a suffix of L0 files (and their associated blob files) are missing,  a valid Version is also available to be saved and recover to. Note that we don't do this if the column family was atomically flushed. Because atomic flush also need a consistent view across the column families, we cannot guarantee that if we are recovering to incomplete version.

Test plan:
Existing tests and added unit tests.

### CODE REVIEW

**Review by pdillinger — APPROVED** [review]
I see VersionBuilder is an almost negligible CPU consumer, so not concerned about that.

LGTM except perhaps some corner cases with blob files. Not a big deal. This seems pretty low risk for disrupting non-best-efforts functionality, and best-efforts itself is reasonably experimental.

I'm not a compaction expert, but the approach seems valid on the face. :)

File: db/version_builder.cc [hunk 1]
```diff
@@ -1188,13 +1256,36 @@ class VersionBuilder::Rep {
         mutable_meta.GetGarbageBlobCount(), mutable_meta.GetGarbageBlobBytes());
   }
 
+  bool BlobFileOnlyLinkedToMissingL0Files(
+      const std::unordered_set<uint64_t>& linked_ssts) const {
+    return !linked_ssts.empty() &&
```
[the remark this answers is not in the document]
Comment (jowlyzhang) [reply db/version_builder.cc]: This is in order to leave the empty case for the regular blob garbage collection logic below. But that is indeed confusing. I have separated it out of the function.

File: db/version_edit_handler.h [hunk 2]
```diff
@@ -245,7 +253,8 @@ class VersionEditHandlerPointInTime : public VersionEditHandler {
       VersionSet* version_set, const std::shared_ptr<IOTracer>& io_tracer,
       const ReadOptions& read_options,
       EpochNumberRequirement epoch_number_requirement =
-          EpochNumberRequirement::kMustPresent);
+          EpochNumberRequirement::kMustPresent,
+      bool allow_incomplete_valid_version = true);
```
[the remark this answers is not in the document]
Comment (jowlyzhang) [reply db/version_edit_handler.h]: I will add some documentation for it in a follow up I'm working with that is related to this set of classes.

File: db/version_builder.cc [hunk 3]
```diff
@@ -1046,6 +1107,13 @@ class VersionBuilder::Rep {
         return s;
       }
     }
+
+    if (track_found_and_missing_files_ && version_updated) {
+      version_updated_since_last_check_ = true;
+      if (!edited_in_atomic_group_ && edit->IsInAtomicGroup()) {
+        edited_in_atomic_group_ = true;
```
Comment (anand1976) [inline db/version_builder.cc:1114]: IIRC, Zorro sets atomic_flush to true, even though it uses only one column family. Maybe we should disable this feature only if # of column families > 1. Longer term, we might want to make this work for atomic flush also.

**Review by anand1976 — APPROVED** [review]
LGTM

File: db/version_builder.cc [hunk 4]
```diff
@@ -1397,8 +1507,108 @@ class VersionBuilder::Rep {
...
+    for (auto iter = expected_sorted_l0_files.begin();
+         iter != expected_sorted_l0_files.end(); iter++) {
+      uint64_t file_number = (*iter)->fd.GetNumber();
+      if (l0_missing_files_.find(file_number) != l0_missing_files_.end()) {
+        assert(unaddressed_missing_files.find(file_number) !=
+               unaddressed_missing_files.end());
+        unaddressed_missing_files.erase(file_number);
+      } else if (!unaddressed_missing_files.empty()) {
+        return false;
+      }
```
[the remark this answers is not in the document]
Comment (jowlyzhang) [reply db/version_builder.cc:1575]: Good catch, thanks!

File: db/version_builder.cc (hunk 3 above)
Comment (jowlyzhang) [reply db/version_builder.cc:1114]: Yes, we can follow up on this and figure a solution for atomic flush too since that could be ZippyDb's future needs.

**Pull request closed without merging**